Jellyfin.Server.Implementations/MediaSegments/MediaSegmentsManager.cs

```diff
@@ -0,0 +1,213 @@
+    /// </summary>
+    /// <param name="segment">The segment to validate.</param>
+    private void ValidateSegment(MediaSegment segment)
+    {
+        if (segment.ItemId.Equals(default))
+        {
+            throw new ArgumentException($"itemId is default: itemId={segment.ItemId} for segment with type '{segment.Type}.{segment.TypeIndex}'");
+        }
+
+        if (segment.StartTicks >= segment.EndTicks)
+        {
+            throw new ArgumentException($"start >= end: {segment.StartTicks}>={segment.EndTicks} for segment itemId '{segment.ItemId}' with type '{segment.Type}.{segment.TypeIndex}'");
+        }
+    }
+
+    /// <summary>
+    /// Delete all segments when itemid is deleted from library.
+    /// </summary>
+    /// <param name="sender">The sending entity.</param>
+    /// <param name="itemChangeEventArgs">The <see cref="ItemChangeEventArgs"/>.</param>
+    private async void LibraryManagerItemRemoved(object? sender, ItemChangeEventArgs itemChangeEventArgs)
+    {
+        await DeleteSegments(itemChangeEventArgs.Item.Id).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async Task<IReadOnlyList<MediaSegment>> DeleteSegments(Guid itemId, int? streamIndex = null, int? typeIndex = null, MediaSegmentType? type = null)
+    {
+        List<MediaSegment> allSegments;
+
+        if (itemId.IsEmpty())
+        {
+            throw new ArgumentException("Default value provided", nameof(itemId));
+        }
+
+        var dbContext = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
+        await using (dbContext.ConfigureAwait(false))
+        {
+            IQueryable<MediaSegment> queryable = dbContext.Segments.Where(s => s.ItemId.Equals(itemId));
+
+            if (streamIndex is not null)
+            {
+                queryable = queryable.Where(s => s.StreamIndex == streamIndex);
+            }
+
+            if (type is not null)
+            {
+                queryable = queryable.Where(s => s.Type == type);
+            }
+
+            if (typeIndex is not null)
+            {
+                queryable = queryable.Where(s => s.TypeIndex == typeIndex);
+            }
+
+            allSegments = await queryable.ToListAsync().ConfigureAwait(false);
+
+            dbContext.Segments.RemoveRange(allSegments);
+            await dbContext.SaveChangesAsync().ConfigureAwait(false);
+        }
+
+        return allSegments;
+    }
+
+    /// <summary>
+    /// Dispose event.
+    /// </summary>
+    /// <param name="disposing">dispose.</param>
+    protected virtual void Dispose(bool disposing)
+    {
+        if (!_disposed)
+        {
+            if (disposing)
+            {
+                _libraryManager.ItemRemoved -= LibraryManagerItemRemoved;
+            }
+
+            _disposed = true;
+        }
+    }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        Dispose(disposing: true);
+        GC.SuppressFinalize(this);
+    }
+}
```

Comment:
Since this is a DI service, you can simplify this by marking this class as sealed and combining the two methods.